firstboot: fix root params with creds and prompting disabled

dbnicholson opened the pull request:
Remove an early return that prevents --prompt-root-password or
--prompt-root-shell and systemd.firstboot=off using credentials. In that case,
arg_prompt_root_password and arg_prompt_root_shell will be false, but the
prompt helpers still need to be called to read the credentials. Furthermore, if
only the root shell has been set, don't overwrite the root password.

dbnicholson commented on src/firstboot/firstboot.c:
```diff
@@ -1144,20 +1136,34 @@ static int process_root_account(int rfd) {
-        if (r < 0)
-                return log_error_errno(r, "Failed to write /etc/shadow: %m");
+        if (hashed_password) {
+                r = write_root_shadow(pfd, hashed_password);
```
Actually, I think you need to run this unconditionally in case it's creating `/etc/shadow` but change `write_root_shadow` to allow a `NULL` password like `write_root_passwd` was changed.

bluca commented:
Please see if some test coverage for these cases can be added to test/units/TEST-74-AUX-UTILS.firstboot.sh

dbnicholson commented:
> Please see if some test coverage for these cases can be added to test/units/TEST-74-AUX-UTILS.firstboot.sh

Took a while to figure out how to run those tests, but I'm making progress several tests that cover my fixes as well as several things that weren't there before. Unfortunately, I can't test the thing I can here to fix. Since the tests run `systemd-firstboot` with `--root`, the `systemd.firstboot` kernel command line parameter is ignored and prompting is never disabled. I could add a `--disable-prompting`/`--credentials-only` CLI option if that's desired.

bluca commented:
The third commit title has a typo

bluca commented:
Does this need some documentation/man updates?

dbnicholson commented:
Fixed the commit title. I don't believe any documentation updates are needed as all the changes are bug fixes to do what the documentation says:

* Make locked and deleted passwords with shadow pointers. I actually just tested this on an older system. If you make the passwd entry without the `x` shadow pointer, then the next time you update with `passwd`, it will leave the encrypted password in `/etc/passwd` for all to see unless you run `pwconv` in the meantime.
* If `/etc/passwd` or `/etc/shadow` exist but don't have a `root` entry, then none of the root options do anything.
* If you run with prompting disabled via `systemd.firstboot=no`, then none of the prompted root options do anything.
* If you force a root shell without also providing a root password, then it removes an existing password.

dbnicholson commented:
I was looking through the failures and I don't think any of them are related to this work.

mkosi / ci (fedora, rawhide, address,undefined, 1, -Og, yes):

```
17/92 systemd:integration-tests / TEST-85-NETWORK-NetworkdNetworkTests                          FAIL            207.57s   exit status 1
```

Seems networkd related.

mkosi / ci (opensuse, tumbleweed, 0, -Og, no)

```
24/92 systemd:integration-tests / TEST-85-NETWORK-NetworkctlTests                               TIMEOUT        1800.16s   killed by signal 15 SIGTERM
27/92 systemd:integration-tests / TEST-85-NETWORK-NetworkdNetworkTests                          TIMEOUT        1800.10s   killed by signal 15 SIGTERM
28/92 systemd:integration-tests / TEST-85-NETWORK-NetworkdRATests                               TIMEOUT        1800.10s   killed by signal 15 SIGTERM
40/92 systemd:integration-tests / TEST-85-NETWORK-NetworkdDHCPClientTests                       TIMEOUT        1800.19s   killed by signal 15 SIGTERM
47/92 systemd:integration-tests / TEST-85-NETWORK-NetworkdDHCPPDTests                           TIMEOUT        1800.15s   killed by signal 15 SIGTERM
92/92 systemd:integration-tests / TEST-13-NSPAWN                                                TIMEOUT        1800.53s   killed by signal 15 SIGTERM
```

Seems mostly networkd related. I'm not sure what's up with the nspawn one, but the last command before the timeout is `importctl`, which seems unrelated.

noble-amd64

```
 8/92 systemd:integration-tests / TEST-64-UDEV-STORAGE-nvme_basic                               FAIL              5.77s   (exit status 241 or signal 113 SIGinvalid)
28/92 systemd:integration-tests / TEST-85-NETWORK-NetworkdDHCPServerRelayAgentTests             FAIL              7.87s   (exit status 241 or signal 113 SIGinvalid)
```

These seem unrelated.

testing-farm:fedora-rawhide-x86_64

```
TEST-46-HOMED:                      FAIL     ( 10 s)
TEST-74-AUX-UTILS:                  FAIL     ( 14 s)
```

These seemed interesting since TEST-74-AUX-UTILS is what I was editing. Similarly, homed is semi-related to firstboot. However, both of them seem to be related to ssh setup:

```
[ 3088.958046] TEST-46-HOMED.sh[796]: + tr -d '\r'
[ 3088.958288] TEST-46-HOMED.sh[795]: + ssh -t -t -4 -p 4711 -i /tmp/homed.id_ecdsa -o 'SetEnv PASSWORD=hunter4711' -o 'StrictHostKeyChecking no' homedsshtest@localhost echo zzz
[ 3088.958496] TEST-46-HOMED.sh[797]: + tee /tmp/homedsshtest.out
[ 3088.984077] TEST-46-HOMED.sh[795]: kex_exchange_identification: read: Connection reset by peer
[ 3088.984077] TEST-46-HOMED.sh[795]: Connection reset by 127.0.0.1 port 4711
```
and
```
[ 4378.615895] TEST-74-AUX-UTILS.sh[1742]: + ssh -o StrictHostKeyChecking=no -v -i /tmp/tmp.Ohw2VZ540j .host cat /etc/machine-id
[ 4378.616211] TEST-74-AUX-UTILS.sh[1743]: + cmp - /etc/machine-id
[ 4378.620093] TEST-74-AUX-UTILS.sh[1742]: OpenSSH_9.8p1, OpenSSL 3.2.2 4 Jun 2024
...
[ 4378.646644] sshd[1745]: fatal: /usr/libexec/openssh/sshd-session does not exist or is not executable
```

Do you see anything that I should try to fix here?

bluca commented:
These are all known issues and unrelated
